src/types/closure.md

```diff
@@ -42,48 +50,148 @@ so that the call to `f` works as if it were:
+
+```rust=
+let s = String::new("S");
+let t = (s, String::new("T"));
+let mut u = (t, String::new("U"));
+
+let c = || {
+    println!("{:?}", u); // u captured by ImmBorrow
+    u.0.truncate(0); // u.0 captured by MutBorrow
+    move_value(u.0.0); // u.0.0 captured by ByValue
+};
+```
+
+Overall the closure will capture `u` by `ByValue`.
+
+### Wild Card Patterns
+Closures only capture data that needs to be read, which means the following closures will not capture `x`
 
 ```rust
-# use std::collections::HashSet;
-#
-struct SetVec {
-    set: HashSet<u32>,
-    vec: Vec<u32>
-}
+let x = 10;
+let c = || {
+    let _ = x;
+};
+
+let c = || match x {
+    _ => println!("Hello World!")
+};
+```
 
-impl SetVec {
-    fn populate(&mut self) {
-        let vec = &mut self.vec;
-        self.set.iter().for_each(|&n| {
-            vec.push(n);
-        })
-    }
-}
+### Capturing references in move contexts
+
+Rust doesn't allow moving fields out references as a result in the case of move closures, when the closure takes ownership of a path, it will only take ownership of a prefix of path that runs up until the first dereference (this may be the entire path if there are no dereferences in the path).
+
+```rust
+struct T(String, String);
+
+let mut t = T(String::from("foo"), String::from("bar"));
+let t = &mut t;
+let c = move || t.0.truncate(0); // closure captures (&mut t.0)
 ```
 
-If, instead, the closure were to use `self.vec` directly, then it would attempt
-to capture `self` by mutable reference. But since `self.set` is already
-borrowed to iterate over, the code would not compile.
+### Raw pointer dereference
+In Rust it's `unsafe` to dereference a raw pointer therefore closures will only capture the prefix of a path that runs upto but not including the first dereference of a raw pointer.
+
+```rust,
+struct T(String, String);
+
+let foo = T(String::from("foo"), String::from("bar"));
+let foo = &foo as *const T;
+
+let c = || unsafe {
+    println!("{}", (*foo).0); // closure captures t
+};
+```
+
+### Reference into unaligned `struct`s
+
+In Rust it's `unsafe` to hold references to unaligned fields in a structure and therefore closures will only capture the prefix of the path that runs upto but not including the first field access into an unaligned struct.
+
+```rust
+#[repr(packed)]
+struct T(String, String);
+
+let t = T(String::from("foo"), String::from("bar"));
+let c = || unsafe {
+    println!("{}", t.0); // closure captures t
+};
+```
+
+
+### `Box` vs other `Deref` implementations
+
+The compiler special cases the implementation of the Deref trait for `Box` and the feature treats `Box` as a special entity as well.
+
+To draw a comparison we look at explamples involving `Rc` and `Box`. The `*rc` is desugared to a call to the trait method `deref` method defined on `Rc`, but `*box` is special cased by the compiler. This implies that the compiler is able to do precise capture on contents of the Box.
```

Comment:
```suggestion
For example, let us look at examples involving `Rc` and `Box`. The `*rc` is desugared to a call to the trait method `deref` defined on `Rc`, but since `*box` is treated differently by the compiler, the compiler is able to do precise capture on contents of the `Box`.
```